src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/TagSet.cs

```diff
@@ -0,0 +1,235 @@
+    }
+
+    public string GetSinglePrechecked() => (string)_tagOrTags!; // we expect this to fail if used on incorrect types
+    public Span<string> GetSpanPrechecked() => (string[])_tagOrTags!; // we expect this to fail if used on incorrect types
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1065:Do not raise exceptions in unexpected locations", Justification = "Intentional; should not be used")]
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Blocker Code Smell", "S3877:Exceptions should not be thrown from unexpected methods", Justification = "Intentional; should not be used")]
+    public override bool Equals(object? obj) => throw new NotSupportedException();
+
+    // [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1065:Do not raise exceptions in unexpected locations", Justification = "Intentional; should not be used")]
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Blocker Code Smell", "S3877:Exceptions should not be thrown from unexpected methods", Justification = "Intentional; should not be used")]
+    public override int GetHashCode() => throw new NotSupportedException();
+
+    public override string ToString() => _tagOrTags switch
+    {
+        string tag => tag,
+        string[] tags => string.Join(", ", tags),
+        _ => "(no tags)",
+    };
+
+    public bool IsEmpty => _tagOrTags is null;
+
+    public int Count => _tagOrTags switch
+    {
+        null => 0,
+        string => 1,
+        string[] arr => arr.Length,
+        _ => 0, // should never happen, but treat as empty
+    };
+
+    internal bool IsArray => _tagOrTags is string[];
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CA2201:Do not raise reserved exception types", Justification = "This is the most appropriate exception here.")]
+    public string this[int index] => _tagOrTags switch
+    {
+        string tag when index == 0 => tag,
+        string[] tags => tags[index],
+        _ => throw new IndexOutOfRangeException(nameof(index)),
+    };
+
+    public void CopyTo(Span<string> target)
+    {
+        switch (_tagOrTags)
+        {
+            case string tag:
+                target[0] = tag;
+                break;
+            case string[] tags:
+                tags.CopyTo(target);
+                break;
+        }
+    }
+
+    internal static TagSet Create(IEnumerable<string>? tags)
+    {
+        if (tags is null)
+        {
+            return Empty;
+        }
+
+        // note that in multi-tag scenarios we always create a defensive copy
+        if (tags is ICollection<string> collection)
+        {
+            switch (collection.Count)
+            {
+                case 0:
+                    return Empty;
+                case 1 when collection is IList<string> list:
+                    return new TagSet(list[0]);
+                case 1:
+                    // avoid the GetEnumerator() alloc
+                    var arr = ArrayPool<string>.Shared.Rent(1);
```

Comment:
For my own knowledge, your opinion on this vs. a ThreadStatic string[1]? (it's even possible that ArrayPool does that internally)

Reply:
Given that there's no chance of re-entrancy, that would be a very valid implementation, and probably a lot less mess than arraypool - which IIRC is already going to be doing some thread-bucketing; I'll go ahead and switch over to that - see a5b816eed83d505af43e89fc3393c4ca0a54d109